libtransmission/rpc-server.cc

```diff
@@ -544,6 +588,108 @@ static auto constexpr ServerStartRetryDelayIncrement = int{ 5 };
 static auto constexpr ServerStartRetryDelayStep = int{ 3 };
 static auto constexpr ServerStartRetryMaxDelay = int{ 60 };
 
+static char const* tr_rpc_address_to_string(tr_rpc_address const* addr, char* buf, size_t buflen)
+{
+    TR_ASSERT(tr_rpc_address_is_valid(addr));
+
+    switch (addr->type)
+    {
+    case TR_RPC_AF_INET:
+        return evutil_inet_ntop(AF_INET, &addr->addr, buf, buflen);
+
+    case TR_RPC_AF_INET6:
+        return evutil_inet_ntop(AF_INET6, &addr->addr, buf, buflen);
+
+    case TR_RPC_AF_UNIX:
+        tr_strlcpy(buf, addr->addr.unixSocketPath, buflen);
+        return buf;
+
+    default:
+        return nullptr;
+    }
+}
+
+static std::string tr_rpc_address_with_port(tr_rpc_server const* server)
+{
+    tr_rpc_address const* addr = server->bindAddress;
+
+    char addr_buf[TrUnixAddrStrLen];
+    tr_rpc_address_to_string(addr, addr_buf, sizeof(addr_buf));
+
+    std::string addr_port_str{ addr_buf };
+    if (addr->type != TR_RPC_AF_UNIX)
+    {
+        addr_port_str.append(":" + std::to_string(tr_rpcGetPort(server)));
+    }
+    return addr_port_str;
+}
+
+static bool tr_rpc_address_from_string(tr_rpc_address* dst, std::string_view src)
+{
+    if (tr_strvStartsWith(src, TrUnixSocketPrefix))
+    {
+        if (src.length() >= TrUnixAddrStrLen)
+        {
+            tr_logAddNamedError(
+                MyName,
+                _("Unix socket path must be fewer than %zu characters (including \"%s\" prefix)"),
+                TrUnixAddrStrLen - 1,
+                TrUnixSocketPrefix);
+            return false;
+        }
+
+        dst->type = TR_RPC_AF_UNIX;
+        tr_strlcpy(dst->addr.unixSocketPath, std::string{ src }.c_str(), TrUnixAddrStrLen);
+        return true;
+    }
+
+    if (evutil_inet_pton(AF_INET, std::string{ src }.c_str(), &dst->addr) == 1)
+    {
+        dst->type = TR_RPC_AF_INET;
+        return true;
+    }
+
+    if (evutil_inet_pton(AF_INET6, std::string{ src }.c_str(), &dst->addr) == 1)
+    {
+        dst->type = TR_RPC_AF_INET6;
+        return true;
+    }
+
+    return false;
+}
+
+static bool bindUnixSocket(struct event_base* base, struct evhttp* httpd, char const* path, int socket_mode)
+{
+#ifdef _WIN32
+    tr_logAddNamedError(
+        MyName,
+        _("Unix sockets are not supported on Windows. Please change \"%s\" in your configuration file."),
+        TR_KEY_rpc_bind_address);
+    return false;
+#else
+    struct sockaddr_un addr;
+    addr.sun_family = AF_UNIX;
+    tr_strlcpy(addr.sun_path, path + strlen(TrUnixSocketPrefix), sizeof(addr.sun_path));
```

Comment:
Here and in one more place, the length of `TrUnixSocketPrefix` is statically known. Maybe turn it into a `std::string_view`?..

Reply:
Do you mean `path`? Otherwise what is the purpose of turning `TrUnixSocketPrefix` into a string_view? Doesn't the compiler optimize this out to the correct value?

Reply:
I'm talking about `TrUnixSocketPrefix`. In a few places, you used `sizeof()` to calculate its size, here (and in one more place) you're calculating it at runtime (via `strlen()`) which is not really necessary. Switching to `std::string_view` will allow to use `std::size()` which won't have any runtime overhead.